### PR TITLE
Geometry Support

### PR DESCRIPTION
**Pull request by mikeringrose**
This was originally based on https://github.com/sequelize/sequelize/pull/3641, however, I fixed some bugs and hopefully made the code more extensible. If you're happy with the path that I am on, I'll add support for MySQL (and more tests).

### CODE REVIEW

File: lib/dialects/abstract/query-generator.js [hunk 1]
```diff
@@ -2245,6 +2248,15 @@ module.exports = (function() {
...
       return value;
+    },
+
+    geometrySelect: function(column) {
+      // just assuming default behavior will return geo json
+      return this.quoteIdentifiers(column);
+    },
+
+    geometryToGeoJson: function(value) {
+      return JSON.parse(value);
```
Comment (janmeier) [inline lib/dialects/abstract/query-generator.js]: This should use WKT righT?

Comment (janmeier) [inline lib/dialects/abstract/query-generator.js]: Not sure parsing belongs in the QG. Perhaps in dialects/X/geojson.js? With a parse and stringify function similar to how PG range works https://github.com/sequelize/sequelize/blob/master/lib/dialects/postgres/range.js

File: lib/data-types.js [hunk 2]
```diff
@@ -526,6 +526,36 @@ ARRAY.is = function(obj, type) {
...
+
+  if (!(this instanceof GEOMETRY)) return new GEOMETRY(options);
+
+  this.options = options;
+  this.type = options.type;
+  this.srid = options.srid;
+};
+util.inherits(GEOMETRY, ABSTRACT);
+
+GEOMETRY.prototype.key = GEOMETRY.key = 'GEOMETRY';
```
Comment (janmeier) [inline lib/data-types.js]: These type defs should be moved to postgres datatypes, and then the general datatypes should be WKT

**Comment by janmeier** [comment]
Looking good so far! Give me shout when you have something more, or if you have issues with WKT support :)

File: lib/data-types.js (hunk 2 above)
Comment (mikeringrose) [inline lib/data-types.js]: My original though is that maybe the parsing should be done in the GEOMETRY
data type object. That way the code is in an obvious place. I'll check out
the range object.

Mike Ringrose

Comment (janmeier) [inline lib/data-types.js]: That's also not a bad idea, but one problem of that could be that the data-types files becomes very long, especially for PG which has a fair amount of specific parsing code (hstore, range)

Letting the datatypes parse values is very close to our idea of the [Datatypes up/down system](https://github.com/sequelize/sequelize/issues/2042) (not implemented yet), so you're on the right track, just ahead of your time :P

**Comment by janmeier** [comment]
Any news @mikeringrose

**Comment by mikeringrose** [comment]
jan,

here's the issue that i'm currently running into. when sequelize performs
an insert the following is execute:

  INSERT INTO "tablename" ("id","geom") VALUES
(DEFAULT,ST_GeomFromGeoJSON('{"type":"Point","coordinates":[34,-72]}'))
RETURNING *;

this causes an issue since we're not generating the select clause here, the
value being returned from the postgres is hex encoded wkb. there are two
solutions, 1) decode the value using hex encoded wkb and then transform to
geojson or 2) generate a proper select statement. i think the latter would
be better moving forward and the former should be now to add this feature.
Any opinions?

Mike Ringrose

On Sun, Jun 14, 2015 at 8:36 AM, Jan Aagaard Meier <notifications@github.com

**Comment by janmeier** [comment]
@mikeringrose Agreed, the right solution would definitely be to generate the correct select statement.

However, I can fully understand that you don't want to undertake that now - and since this is the only case where this has been a problem up to now it would be a lot of work just for this feature :). We would be fine with merging 1 for now.

**Comment by mikeringrose** [comment]
Awesome. Thanks Jan. I update my PR in the morning and then I wouldn't mind
helping out on some other features. Sorry for the delay, lots of projects,
little time. :)

Mike Ringrose

**Comment by mikeringrose** [comment]
I added a few tests, moved the parsing out of query-generator, and i'm now using the raw value returned from postgresql instead of converting it to GeoJSON. The new way I read the value in as binary, then as WKB, and finally to GeoJSON.

File: logfile [hunk 3]
```diff
@@ -0,0 +1,90 @@
+LOG:  database system was shut down at 2015-06-25 07:14:13 EDT
```
Comment (janmeier) [inline logfile]: :fire:

**Comment by janmeier** [comment]
@mikeringrose Sounds very good! If everything is passing locally for you you should rebase so we can run the tests on travis :)

File: test/unit/sql/data-types.test.js [hunk 4]
```diff
@@ -282,6 +282,33 @@ suite(Support.getTestDialectTeaser('SQL'), function() {
...
+        postgres: 'GEOMETRY(LINESTRING)',
+      });
+
+      testsql('GEOMETRY(\'POLYGON\')', DataTypes.GEOMETRY('POLYGON'), {
+        default: 'GEOMETRY(POLYGON)',
+        postgres: 'GEOMETRY(POLYGON)'
+      });
+
+      testsql('GEOMETRY(\'POINT\',4326)', DataTypes.GEOMETRY('POINT', 4326), {
+        default: 'GEOMETRY(POINT,4326)',
```
Comment (janmeier) [inline test/unit/sql/data-types.test.js]: I believe this should be just `POINT` in the general case (for [mysql](https://dev.mysql.com/doc/refman/5.0/en/spatial-datatypes.html) at least

File: lib/dialects/postgres/query-generator.js [hunk 5]
```diff
@@ -893,6 +893,13 @@ module.exports = (function() {
         return 'ARRAY[' + value.map(function (v) {
           return SqlString.escape(JSON.stringify(v), false, this.options.timezone, this.dialect, field);
         }, this).join(',') + ']::JSON[]';
+      } else if (value && field && field.type instanceof DataTypes.GEOMETRY) {
+        var s = "ST_GeomFromGeoJSON('" + JSON.stringify(value) + "'";
```
Comment (janmeier) [inline lib/dialects/postgres/query-generator.js]: Still missing the equivalent of this code for the other dialects, correct? Which converts geojson to wkt before inserting it

File: test/integration/dialects/postgres/dao.test.js [hunk 6]
```diff
@@ -867,6 +868,35 @@ if (dialect.match(/^postgres/)) {
       });
     });
 
+    it('should save geometry correctly', function() {
+      var point = { type: 'Point', coordinates: [39.807222,-76.984722] };
+      return this.User.create({ username: 'user', email: ['foo@bar.com'], location: point}).then(function(newUser) {
+        expect(point).to.eql(newUser.location);
```
Comment (janmeier) [inline test/integration/dialects/postgres/dao.test.js]: Theoretically some internals might mutate `.location`, which we wouldn't notice here. Its better to say `expect({ type: 'Point', coordinates: [39.807222,-76.984722] }).to.deep.eql(newUser.location);`, even though its duplication :)

**Comment by mikeringrose** [comment]
Cool. I'll make that change. I was thinking that since this PR has been going on for a while, I'd add the postgis implementation first followed by the msyql implementation. I have a lot more experience with postgis but a mysql implemenation should follow fairly fast.

File: lib/instance.js [hunk 7]
```diff
@@ -383,6 +383,13 @@ module.exports = (function() {
             }
           }
 
+          // parse the well known text
+          if (this.Model._hasGeometryAttributes && this.Model._isGeometryAttribute(key) && value !== null && value !== undefined && !value._isSequelizeMethod) {
```
Comment (janmeier) [inline lib/instance.js]: We don't need the ` && value !== null && value !== undefined && !value._isSequelizeMethod` part, when we check `isString` right after :)

File: lib/dialects/abstract/query-generator.js [hunk 8]
```diff
@@ -2245,6 +2248,10 @@ module.exports = (function() {
 
     booleanValue: function(value) {
       return value;
+    },
+
+    geometrySelect: function(column) {
```
Comment (janmeier) [inline lib/dialects/abstract/query-generator.js]: This function is never overridden by dialects, so it could just be inlined, right?

**Comment by janmeier** [comment]
Very clean and readable code @mikeringrose - good work so far!

**Comment by mikeringrose** [comment]
Are you cool if I just re-branch off of the latest code and apply my changes? I'm running into some painful merge conflicts.

**Comment by janmeier** [comment]
Yep, thats fine

File: lib/instance.js (hunk 7 above)
Comment (mikeringrose) [inline lib/instance.js]: Good catch. Will change.

**Comment by mikeringrose** [comment]
All code has been moved to the following PR: https://github.com/sequelize/sequelize/pull/4013.

**Pull request closed without merging**